File: 94-binary-tree-inorder-traversal/solution.cpp
#include <vector>
#include <deque>

struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
  std::vector<int> inorderTraversal(TreeNode *root)
  {
    std::vector<int> result;
    std::deque<const TreeNode *> work_deque;
    const TreeNode *curr = root;
    while (curr != nullptr || !work_deque.empty())
    {
      while (curr != nullptr)
      {
        work_deque.emplace_back(curr);
        curr = curr->left;
      }

      const TreeNode *const node = work_deque.back();
      work_deque.pop_back();
      result.emplace_back(node->val);
      curr = node->right;
    }

    return result;
  }
};

File: 94-binary-tree-inorder-traversal/solution.py
from typing import Optional
from collections import deque


class TreeNode:
  def __init__(self, val=0, left: Optional['TreeNode'] = None, right: Optional['TreeNode'] = None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def inorderTraversal(self, root: Optional[TreeNode]) -> list[int]:
    result = []
    work_deque: deque[TreeNode] = deque()
    curr = root
    while curr is not None or work_deque:
      if curr is not None:
        work_deque.append(curr)
        curr = curr.left
        continue

      curr = work_deque.pop()
      result.append(curr.val)
      curr = curr.right

    return result
